WIP: Sklearn kdtree

kain88-de opened the pull request:
Fixes #383 

I have completely removed the old KDTree library. Functions that have used the old class
directly have been adopted to use the scikit-learn implementation. The `AtomNeighborSearch` class
was a nice abstraction that worked directly with `AtomGroups`, it is still available but internally it
uses the new KDTree implementation. I tried to make the commits self contained which should make it easier to review this.

The tests have been updated for the new `AtomNeighborSearch` API.

TODO:
=====
- [x] write Doc strings for `AtomNeighborSearch`
- [x] add removal of NeighborSearch to changelog
- [x] deprecate KDTree completely with notice to use scikit-learn instead
- [x] fix deprecation test failure

kain88-de commented:
Btw what are is the purpose of the depcreation test that is failing? Should it test that depcreated imports are still available? In that case I would remove that specific test because I have removed the module. `AtomNeighborSearch` now lives in it's own module, where I'm thinking if I should move it to the analysis sub-module since it is only used there.

richardjgowers commented:
I think the failing deprecation test is that KDTree used to live in `MDAnalysis.KDTree` and we moved it to `MDAnalysis.lib.KDTree`, but we want the old address to work (for now), and issue a deprecation warning.  So there's a stub module which issues the warning then imports the other module.

kain88-de commented:
I guess then I could readd the stub module and load the new `atomneighbors` instead and issue a
warning about the move and rename. I think I will also move the `atomneighbors` module into `MDAnalysis.analysis` since it is only used there in MDAnalysis and I figure it is something people want to use it in their analysis.

richardjgowers commented:
The submodule .analysis is complete tools which can have MDA dependencies,
whilst .lib is building blocks without MDA dependencies. I need to properly
catch up on this to figure out where this should live.

On 1:24PM, Wed, 19 Aug 2015 kain88-de <notifications@github.com> wrote:

> I guess then I could readd the stub module and load the new atomneighbors
> instead and issue a
> warning about the move and rename. I think I will also move the
> atomneighbors module into MDAnalysis.analysis since it is only used there
> in MDAnalysis and I figure it is something people want to use it in their
> analysis.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/395#issuecomment-132572165>
> .
>

kain88-de commented:
Once you found out where it should live I will finish my TODO-list to have this mergeable.

Btw it is ok that I removed the `NeighborSearch` class? I didn't see any particular value for a
KDTree wrapper in MDAnalysis and I figured if people want to use a KDTree on arbitrary coordinates
it might be better if they use one of the common ML-python packages

kain88-de commented:
Tests are all passing now and I fixed the doc string for the atomneihgborsearch class. The I didn't test the density analysis because a unit-test for it doesn't exists.

Anything else missing?

richardjgowers commented on package/MDAnalysis/lib/NeighborSearch.py:
```diff
@@ -0,0 +1,75 @@
+import numpy
```
This file will need a header comment for the rst docs won't it?

kain88-de replied:
I added a header comment. I'm not very familiar with the rst docs. If I missed defining some rst specific stuff please tell me. I would also appreciate a link to the appropriate sphinx documentation for this.

richardjgowers commented on package/MDAnalysis/lib/__init__.py:
```diff
@@ -20,10 +20,9 @@
 ================================================================
 """
 
-__all__ = ['KDTree', 'log', 'transformations', 'util', 'mdamath', 'distances',
```
Add NeighborSearch here?

richardjgowers commented:
Ok header looks fine.  What needs doing now is for the .rst stubs to be updated.  If you go to `doc/sphinx/source/documentation_pages/` there is a lot of .rst files that mirror the layout of the package.  You need to update these (KDTree and lib) so that they reflect the new structure.  I think this will be, deleting the KDTree directory, add a new .rst for NeighborSearch (copy another submodule, should just be 2 lines) and add a link to `lib_modules.rst`

Link to links:
https://github.com/MDAnalysis/mdanalysis/wiki/WritingDocumentation#online-docs-format-sphinx-restructuredtext

kain88-de commented:
I removed the old KDTree rst files and added a short one for the new NeighborSearch module. The formatting is different then from the other include in `lib_modules.rst`. But I have no idea why.

Plus I suddenly have all these import error where I have no real idea what might have caused them.

This is an example for a failed import.
https://travis-ci.org/MDAnalysis/mdanalysis/builds/77053307#L459

I assume loading `NeighborSearch` in `lib/__init__.py` somehow has caused circular imports with `AtomGroup`. Removing that import fixes the import-errors I'm seeing. @richardjgowers 
Do you have any idea how to fix this properly? Currently I can only think of using

```python
from ..exceptions import SelectionError, .....
```

in `AtomgGourp.py`

kain88-de commented:
> Plus I suddenly have all these import error where I have no real idea what might have caused them

Yup it was a circular import see 80078feaaf715f1efb8c86d9837d45138ae2f04b.

richardjgowers commented:
Awesome
